### ISSUE TITLE
Issue with FlatKitOutline Image effect on Quest apk

### ISSUE THREAD

**Issue by D1ngoli** [Dustyroom/flat-kit-doc#23]
I have an issue with the global Outline Image affect when using Oculus Quest with URP.
I have set in de Render features the Outline FlatKit component. Within the editor, the effect looks great.
However when playing from an apk, the Quest will only show a grey plane when booting up the project.

When using Oculus Link and playing the project from the Editor the outline effect works, however only at the left eye it is displayed correctly. At the right eye the lines seem to update a couple of seconds too late, which make the lines jump slightly when moving your head, which is very distracting.

**Expected behavior**
It would be nice when the effect would look the same in the editor as when playing in VR on a Quest.
Now both versions in VR with native apk or playing via Oculus link the effect does not behave as expected, even though when playing non VR the effect is displayed correctly.

**Screenshots**
Settings:
![image](https://user-images.githubusercontent.com/80272011/110346192-982f7f80-802f-11eb-9149-d63f57fd0503.png)
![image](https://user-images.githubusercontent.com/80272011/110346294-b2695d80-802f-11eb-99db-78fded814caf.png)
![image](https://user-images.githubusercontent.com/80272011/110346382-c614c400-802f-11eb-8b80-fff59e227163.png)

Apk:
![screenshot-apk](https://user-images.githubusercontent.com/80272011/110347325-cbbed980-8030-11eb-9a36-46fdda15ff78.jpg)

From editor when using Oculus link (right eye, moving head):
![com oculus xrstreamingclient-20210308-171039](https://user-images.githubusercontent.com/80272011/110348003-7a631a00-8031-11eb-9330-67aaa0caefb0.jpg)

**Unity details:**
 - Flat Kit version:  2.1.4
 - Unity version: 2019.4.5f1
 - Graphics Pipeline: URP
 - URP version : 7.3.1
 - Development OS: Windows

**Build platform:**
 - Platform: Android
 - Test device: Oculus Quest 1

**Comment by dustyroom-studio** [comment]
Hi,

Before trying to replicate your issue, I can tell there's one possible reason for your issue: **you've got a too old version of Unity**, which brings a **too old version of Universal RP**, and in addition it possibly didn't let you update/install the newest version of Flat Kit (though it says you've got the latest version).

This first issue may be because you either use Unity version prior to 2019.4.19f1 / 2020.1.2f1 or you work in Built-In Rendering Pipeline, the Water is made for Universal RP only.

There is bug with Unity Package Manager for specific versions. It says you ave got the latest versions of assets but in reality it may nit be the case. So, since you use Unity version prior 2019.4.19f1, where Unity addressed this bug, please do the following:

1. Delete Flat Kit from Project Tab in Unity,
2. Close Unity,
3. Clear the Package Manager cache. You can find the cache here:
C:\Users\accountName\AppData\Roaming\Unity\Asset Store-5.x
4. Install Unity at least 2019.4.19f1 or, better, a later one — if you use 2019's. If you are planning to use 2020's — please install at least 2020.2.1 or, better, a later one.
5. Create / open a Universal RP project with the new version of Unity.
5. Go to Window -> Package Manager -> select 'My Assets' in 'Packages' drop down menu -> select Flat Kit -> press 'Download' -> after downloading press 'Import'.
6. Install Flat Kit as described in [the installation guide here](https://flatkit.dustyroom.com/#2-quick-start-beginning-to-work-with-flat-kit). It is important that you completed all the steps.

If after updating you'll still have this issue, we'll dig into it.

**Comment by D1ngoli** [comment]
Firstly, thank you for the fast reply!

When using Unity 2019.4.5f1 the package manager really did not let me update the asset and was stuck with the old version of Flat Kit. 
And then I followed your steps.

0. Deleted FlatKit from the Package cache
1. Installed Unity 2019.4.21f1
2. Made a clean project with URP (to be sure)
3. Opened Package Manager and downloaded Flat Kit version 2.1.4

Everything works at first glance. As a test i used the Wanderer example scene with the Outline image effect. 
However when I was building an apk I got shader errors:
![image](https://user-images.githubusercontent.com/80272011/110630530-bfa85880-81a5-11eb-9882-8ff6bfefcfe9.png)
After building the Image effect would stop being displayed in the Editor and the shader, FlatKit/Stylized Surface is now at the 'failed to compile' shaders.
When updating URP to newer and older versions the shader error after making a build still remains, however the Outline effect is displayed again when switching versions of URP.

The weird thing is, that the shaders and effect works on the Quest without any issues. 
**The Outline image effect does work now!**
Only when running the scene with the Oculus link, the issue with the right eye however still remains.

Also when applying the new FlatKit shader in my old project in Unity 2019.4.5f1, i get the same shader errors and in a new build, all gameobjects with a material with the FlatKit shader are invisible.

So only if the shader errors can be fixed, everything would work as expected.
I still love this asset but I'm a little stuck now.

**Unity details:**
Flat Kit version: 2.1.4
Unity version: 2019.4.12f1
Graphics Pipeline: URP
URP version : 7.5.3
Development OS: Windows

**Build platform:**
Platform: Android
Test device: Oculus Quest 1

**Comment by dustyroom-studio** [comment]
Hi,

Could you please verify a few things:
You have done the following:
- Set Edit -> Project Settings -> Graphics -> Scriptable Render Pipeline to "[FlatKit] Example Settings URP"
- Set Project Settings -> Quality Settings -> Rendering to "[FlatKit] Example Settings URP" (for all quality settings)
- You have tried re-launching Unity

Is there an opportunity that you would try to install Unity 2020.2.x (currently it's 2020.2.7f1) and see it the errors are gone?
Unity 2020.2.x has a newer version of Universal RP, which should fix the issue.

Please let me know how it goes.

**Comment by D1ngoli** [comment]
Hi

- The Renderpipeline is set to "[FlatKit] Example Settings URP" in Graphics
- The Renderpipeline is set to "[FlatKit] Example Settings URP" in Quality Settings
- i relaucnhed Unity

Screenshots:
![image](https://user-images.githubusercontent.com/80272011/110662665-0148fb80-81c6-11eb-81cc-021729b2e17f.png)
![image](https://user-images.githubusercontent.com/80272011/110662721-0f971780-81c6-11eb-89b1-ea57c7415e5c.png)

when building:
![image](https://user-images.githubusercontent.com/80272011/110662783-20e02400-81c6-11eb-8c97-1e305c914414.png)

To upgrade my project to Unity 2020.2.x would be a real big hassle because it is a big project with a lot of plugins.
It would be really nice if it would be ideally possible to get it in Unity 2019.4,5f1 running.

**Comment by dustyroom-studio** [comment]
The latest version of Flat Kit supports Unity 2019.4.5f1, but it seems the version you are using is not the latest. This may be due to a bug in Unity which [we've described here](https://github.com/Dustyroom/flat-kit-doc/blob/master/index.md#q-after-importingupdating-flat-kit-the-shaders-failed-to-compile-x-shader-is-missing-from-the-list-why).

Are you seeing a Readme file in the root of the Flat Kit folder? That file should have the Flat Kit version (if you have the latest).

**Comment by D1ngoli** [comment]
Hi,

I have the readme file and it saying that i have Unity 2020.2.6f1 instead of 2019.4.5f1.
I have downloaded the asset package with Unity 2019.4,12f1.

This is the debug info it posts:
Flat Kit: 2.1.3, URP: 7.4.3, Unity: 2020.2.6f1

Screenshot:
![image](https://user-images.githubusercontent.com/80272011/110758669-11f08480-824d-11eb-83a4-36d7cc779727.png)
![image](https://user-images.githubusercontent.com/80272011/110776943-dc559680-8260-11eb-9878-40de5a71edc0.png)


**Unity** details:
Flat Kit version: 2.1.3
Unity version: 2019.4.5f1
Graphics Pipeline: URP
URP version : 7.3.1
Development OS: Windows
Build platform:

Platform: Android
Test device: Oculus Quest 1



However i now installed a fresh 2019.4.22f1 Unity version with a complete empty project with the URP renderpipeline. 
I removed beforehand the Asset cache in Asset Store- 5.x and redownloaded it from the Package manager and it shows 2.1.4
![image](https://user-images.githubusercontent.com/80272011/110779650-12484a00-8264-11eb-8eeb-c5e84c4dc45e.png)
 but the readme insists it is 2.1.3.
 
 I opened one of the scenes and added them in the build settings. I made sure that the Renderpipline Settings in Quality and Graphics are set to '[FlatKit] Example Settings URP'
 
 When installing everything from scratch the project works as expected. Only when building with a scene with FlatKit shaders, the shaders are failing to compile.
 
![image](https://user-images.githubusercontent.com/80272011/110779851-51769b00-8264-11eb-8912-0975b38dae51.png)

**Unity** details:
Flat Kit version: 2.1.3?
Unity version: 2019.4.22f1
Graphics Pipeline: URP
URP version : 7.5.3
Development OS: Windows
Build platform:

Platform: Android
Test device: Oculus Quest 1



Sorry for being so annoying but i cannot resolve the issue on my own.

At this point I can try to install Unity 2020 as a test and see if the problem persists in a new project.

**Comment by dustyroom-studio** [comment]
No worries, thanks for helping us debug this.
Can you please copy the full error messages of the two errors you're getting when trying to build?
Also, does it build successfully for any other platform (e.g. Windows)?

**Comment by D1ngoli** [comment]
This is the Console log after building with a fresh start of Unity.
![image](https://user-images.githubusercontent.com/80272011/110794857-796efa00-8276-11eb-9ef4-86c60f307a68.png)

Full error messages:

1. Shader error in 'FlatKit/Stylized Surface': failed to open source file: 'Packages/com.unity.render-pipelines.universal/Shaders/SimpleLitGBufferPass.hlsl' at line 249 (on gles3)

Compiling Vertex program with _MIXED_LIGHTING_SUBTRACTIVE
Platform defines: UNITY_NO_DXT5nm UNITY_NO_RGBM UNITY_ENABLE_REFLECTION_BUFFERS UNITY_FRAMEBUFFER_FETCH_AVAILABLE UNITY_NO_CUBEMAP_ARRAY UNITY_NO_SCREENSPACE_SHADOWS UNITY_PBS_USE_BRDF3 UNITY_NO_FULL_STANDARD_SHADER SHADER_API_MOBILE UNITY_LIGHTMAP_DLDR_ENCODING

2. Shader error in 'FlatKit/Stylized Surface': failed to open source file: 'Packages/com.unity.render-pipelines.universal/Shaders/DepthNormalsPass.hlsl' at line 316 (on gles3)

Compiling Vertex program
Platform defines: UNITY_NO_DXT5nm UNITY_NO_RGBM UNITY_ENABLE_REFLECTION_BUFFERS UNITY_FRAMEBUFFER_FETCH_AVAILABLE UNITY_NO_CUBEMAP_ARRAY UNITY_NO_SCREENSPACE_SHADOWS UNITY_PBS_USE_BRDF3 UNITY_NO_FULL_STANDARD_SHADER SHADER_API_MOBILE UNITY_LIGHTMAP_DLDR_ENCODING

3. Shader error in 'FlatKit/Stylized Surface With Outline': failed to open source file: 'Packages/com.unity.render-pipelines.universal/Shaders/SimpleLitGBufferPass.hlsl' at line 258 (on gles3)

Compiling Vertex program
Platform defines: UNITY_NO_DXT5nm UNITY_NO_RGBM UNITY_ENABLE_REFLECTION_BUFFERS UNITY_FRAMEBUFFER_FETCH_AVAILABLE UNITY_NO_CUBEMAP_ARRAY UNITY_NO_SCREENSPACE_SHADOWS UNITY_PBS_USE_BRDF3 UNITY_NO_FULL_STANDARD_SHADER SHADER_API_MOBILE UNITY_LIGHTMAP_DLDR_ENCODING

4. Shader error in 'FlatKit/Stylized Surface With Outline': failed to open source file: 'Packages/com.unity.render-pipelines.universal/Shaders/DepthNormalsPass.hlsl' at line 325 (on gles3)

Compiling Vertex program with STEREO_MULTIVIEW_ON
Platform defines: UNITY_NO_DXT5nm UNITY_NO_RGBM UNITY_ENABLE_REFLECTION_BUFFERS UNITY_FRAMEBUFFER_FETCH_AVAILABLE UNITY_NO_CUBEMAP_ARRAY UNITY_NO_SCREENSPACE_SHADOWS UNITY_PBS_USE_BRDF3 UNITY_NO_FULL_STANDARD_SHADER SHADER_API_MOBILE UNITY_LIGHTMAP_DLDR_ENCODING




When building on Windows with the Mono scripting backend, no errors are displayed when building.
Within the application the materials show up fine however in the Editor, the shaders are still displayed in the 'failed to compile' section'

In the Editor: 
![image](https://user-images.githubusercontent.com/80272011/110798420-72e28180-827a-11eb-8c30-6a09e43fc2a1.png)
![image](https://user-images.githubusercontent.com/80272011/110798457-7bd35300-827a-11eb-8654-12ad0ac5ceda.png)

**Comment by D1ngoli** [comment]
As an experiment I made a new project in Unity 2020.2.7f1. 
Using the same plugins then before and settings...
no errors!

On Unity 2020.2.7f1 it works as a charm!

Sadly it would take a lot of effort for my team to update our project from Unity 2019.4.5f1 to Unity 2020,2.7f1. 
It's actually not an option for us to upgrade and FlatKit is advertised that it works on Unity 2019.x,x .

However big thanks so far for the support and help.

**Comment by dustyroom-studio** [comment]
Thank you for doing the tests. And yes, I understand, updating the system during an ongoing project is a bad idea.

Flat Kit does work with earlier versions of Unity, but the URP version of Flat Kit heavy relies on Universal RP, which is changed and updated very frequently, sometimes silently. To make Flat Kit to comply with these updates, in other words, to make sure it is somewhat future proof, at this time, we have to develop and test Flat Kit mostly with 2020's versions.

It is quite obvious that your the issue was due to the outdated URP, with the updated URP that comes with 2020.2.xf1 this error is known to be fixed. But, strangely, I still can't reproduce your problem on my 2019's Unity versions, all shaders compile normally.

In other words, it should still work in 2019's. But some features, especially when working with Oculus, are just prone to less errors with Unity 2020.2.7f1, for example.

If you'd like, please send me an email to info@dustyroom.com, and I'll send you the direct link to the latest version of Flat Kit so that you could try importing it directly to try it in your 2019.

Also, I feel the pain of being unavailable to update the project now, so if you can't find the use of Flat Kit now, let me know as well, if you would like a refund for now.

Cheers!

**Comment by r0ckl0bsta** [comment]
Hi there,

I was having this same issue using the FlatKitOutline render effect in URP. My work around for this was to set my Desktop Stereo Rendering Mode to Multi Pass, and disable Shared Depth buffer.

![image](https://user-images.githubusercontent.com/79990644/113494634-1ad51e80-94b8-11eb-9565-5c26552044e7.png)

Hope that helps as a temporary workaround!

**Comment by dustyroom-studio** [comment]
Thanks for the details,
While we work on single-pass support, I'm going to add the multi-pass workaround to the FAQ.

Related issue: https://github.com/Dustyroom/flat-kit-doc/issues/35

**Comment by mutp** [comment]
Hello, I'm facing the same issue with Single Pass XR. Using Unity 2021.1.7f1 and URP 11. 

Do you have an estimated time to fix this issue?

**Comment by dustyroom-studio** [comment]
Thanks for reporting, we got some pointers from Unity and will try to address the issue this week.

**Comment by dustyroom-studio** [comment]
After sinking more hours into this issue, there's still no concrete progress. Unity really needs to release a sample of an image effect working with single-pass XR. Please chime in on [this forum post](https://forum.unity.com/threads/urp-rendererfeatures-and-stereo-cameras.1117447/#post-7369409) if you can.

**Comment by dustyroom-studio** [comment]
Hi! We have added Single Pass Stereo support for Unity 2021 and later in Flat Kit 2.9.0, please check it out.
